sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -468,7 +489,13 @@ case class DataSource(
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
 
-    providingClass.newInstance() match {
+    val resolvedRelation = providingClass.newInstance() match {
+      case relationToCheck: DataSourceValidator =>
```

Comment:
I tried though, I couldn't because the check at this point couldn't pass [this `JDBCWrite` test](https://github.com/apache/spark/pull/17758/files#diff-4e7d97319461efb33f06b6d245254dcdL477). This test catches name duplication specified by a JDBC option `createTableColumnTypes`. But, if we move [this check](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L744) in `JdbcUtils` into here, [the following logic](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L764) eats the duplication and `r.schema` above does not have the duplication (so, I think this is a corner case though, IIUC we couldn't catch at this point for JDBC relations)....